package_test.go

```diff
@@ -1017,41 +1021,204 @@ func (s *PackageSuite) TestTransactions(c *C) {
+		cacheMutex.RLock()
+		defer cacheMutex.RUnlock()
+		dbCache, ok1 := stmtDBCache[sID]
+		var ok2 bool
+		if ok1 {
+			_, ok2 = dbCache[dbID]
+		}
+		_, ok3 := dbStmtCache[dbID][sID]
+		c.Assert(ok2, Equals, inCache)
+		c.Assert(ok3, Equals, inCache)
+	}
+
+	checkDBNotInCache := func(dbID int64) {
+		cacheMutex.RLock()
+		defer cacheMutex.RUnlock()
+		for _, dbCache := range stmtDBCache {
+			_, ok := dbCache[dbID]
+			c.Assert(ok, Equals, false)
+		}
+		_, ok := dbStmtCache[dbID]
+		c.Assert(ok, Equals, false)
+	}
+
+	checkCacheEmpty := func() {
+		cacheMutex.RLock()
+		defer cacheMutex.RUnlock()
+		c.Assert(dbStmtCache, HasLen, 0)
+		c.Assert(stmtDBCache, HasLen, 0)
+	}
+
+	q1 := `
+		SELECT &Person.*
+		FROM person
+		WHERE name = "Fred"
+	`
+	q2 := `
+		SELECT &Person.*
+		FROM person
+		WHERE name = "Mark"
+	`
+
+	// For a Statement or DB to be removed from the cache it needs to go out of
+	// scope and be garbage collected. Because of this, the tests below make
+	// extensive use of functions to "forget" statements and databases.
+
+	// createAndCacheStmt takes a db and prepares a statement on it.
+	createAndCacheStmt := func(db *sqlair.DB) (stmtID int64) {
+		// Create stmt.
+		stmt, err := sqlair.Prepare(q1, Person{})
+		c.Assert(err, IsNil)
+
+		// Start a query with stmt on db.
+		// This will prepare the stmt on the db.
+		c.Assert(db.Query(nil, stmt).Get(&p), IsNil)
+		// Check that it now is in the cache.
+		checkStmtCache(db.CacheID(), stmt.CacheID(), true)
+		return stmt.CacheID()
+	}
+
+	// testStmtsOnDB prepares a given statement on the db then creates a
+	// statement inside another function and checks it has been cleared from
+	// the cache on garbage collection.
+	testStmtsOnDB := func(db *sqlair.DB, stmt *sqlair.Statement) {
+		// Start a query with stmt on db. This will prepare stmt on db.
+		c.Assert(db.Query(nil, stmt).Get(&p), IsNil)
+		// Check that it now is in the cache.
+		checkStmtCache(db.CacheID(), stmt.CacheID(), true)
+
+		funcStmtID := createAndCacheStmt(db)
+		// Run the garbage collector and wait one millisecond for the finalizer to finish.
+		runtime.GC()
+		time.Sleep(1 * time.Millisecond)
+
+		checkStmtCache(db.CacheID(), funcStmtID, false)
+	}
+
+	createDBAndTestStmt := func(stmt *sqlair.Statement) (dbID int64) {
+		// Create db.
+		tables, sqldb, err := personAndAddressDB()
+		c.Assert(err, IsNil)
+		db := sqlair.NewDB(sqldb)
+		defer func() {
+			c.Assert(dropTables(db, tables...), IsNil)
+		}()
+
+		// Test stmt.
+		testStmtsOnDB(db, stmt)
+
+		return db.CacheID()
+	}
+
+	createStmtAndTestOnDBs := func() {
+		// Create stmt.
+		stmt, err := sqlair.Prepare(q2, Person{})
+		c.Assert(err, IsNil)
 
-	err = db.Query(ctx, sqlair.MustPrepare(dropTables)).Run()
+		db1ID := createDBAndTestStmt(stmt)
+		db2ID := createDBAndTestStmt(stmt)
+
+		// Run the garbage collector and wait one millisecond for the finalizer to finish.
+		runtime.GC()
+		time.Sleep(1 * time.Millisecond)
```

Comment:
I'll do a thorough investigation, this was used in the Go runtime package test, but in the latest versions they have actually updated their tests to use channels to indicate when the finalizer is done. We can't do that here because it would involve using a different finalizer to the one in the code so I have stuck with the old approach for now, though I would like to do a more thorough investigation at some point.